prometheus/client.go

```diff
@@ -285,27 +285,39 @@ func (in *Client) GetFlags() (prom_v1.FlagsResult, error) {
 	return flags, nil
 }
 
-// GetMetricsForLabels returns a list of metrics existing for the provided labels set
-func (in *Client) GetMetricsForLabels(labels []string) ([]string, error) {
-	// Arbitrarily set time range. Meaning that discovery works with metrics produced within last hour
-	end := time.Now()
-	start := end.Add(-time.Hour)
-	log.Tracef("[Prom] GetMetricsForLabels: %v", labels)
-	results, warnings, err := in.api.Series(in.ctx, labels, start, end)
+// GetMetricsForLabels returns a list of metrics existing for the provided labels set. Only metrics that match a name in the given
+// list of metricNames will be returned - others will be ignored.
+func (in *Client) GetMetricsForLabels(metricNames []string, labelQueryString string) ([]string, error) {
+	if len(metricNames) == 0 {
+		return []string{}, nil
+	}
+
+	log.Tracef("[Prom] GetMetricsForLabels: labels=[%v] metricNames=[%v]", labelQueryString, metricNames)
+	startT := time.Now()
+	queryString := fmt.Sprintf("count(%v) by (__name__)", labelQueryString)
+	results, warnings, err := in.api.Query(in.ctx, queryString, time.Now())
 	if warnings != nil && len(warnings) > 0 {
 		log.Warningf("GetMetricsForLabels. Prometheus Warnings: [%s]", strings.Join(warnings, ","))
 	}
 	if err != nil {
 		return nil, errors.NewServiceUnavailable(err.Error())
 	}
 
-	var names []string
-	for _, labelSet := range results {
-		if name, ok := labelSet["__name__"]; ok {
-			names = append(names, string(name))
+	metricsWeAreLookingFor := make(map[string]bool, len(metricNames))
+	for i := 0; i < len(metricNames); i++ {
+		metricsWeAreLookingFor[metricNames[i]] = true
+	}
+
+	metricsWeFound := make([]string, 0, 5)
```

Comment:
super-minor but you could give metricsWeFound capacity len(metricNames) (as opposed to 5)

Reply:
The reason I did NOT do that is the majority of those metric names (the vast majority, in fact) will not match. In fact, the number of elements that will be placed inside "metricsWeFound" will usually be 0 or 1 (there are no custom dashboards for the majority of workloads except for "Envoy" for those workloads with sidecars). And for those workloads that DO have custom dashboards, there will usually never be more than 1 or 2 custom dashboards. So even the size of "5" is probably larger than what will be needed. I just put "5" to capture probably every use-case we'd run in to. Even with 5, the vast majority of the time I suspect 4 will be empty.